File: Two Pointers/M-1456. Maximum Number of Vowels in a Substring of Given Length/M-1456. Maximum Number of Vowels in a Substring of Given Length.ts
function maxVowels(s: string, k: number): number {
  let [l, r] = [0, 0];
  let maxVowelCount = 0;
  let vowelCount = 0;
  let vowels = new Set(['a', 'e', 'i', 'o', 'u']);
  while (r < s.length) {
    if (vowels.has(s[r])) {
      vowelCount++;
      maxVowelCount = Math.max(maxVowelCount, vowelCount);
    }

    /** reach to the window size */
    if (r - l === k - 1) {
      if (vowels.has(s[l])) vowelCount--;
      l++;
    }
    r++;
  }
  return maxVowelCount;
}

const testData = [
  {
    s: 'abciiidef',
    k: 3,
    ans: 3,
  },
  {
    s: 'aeiou',
    k: 2,
    ans: 2,
  },
  {
    s: 'leetcode',
    k: 3,
    ans: 2,
  },
];

for (const data of testData) {
  console.log(data['ans'] === maxVowels(data['s'], data['k']));
}

File: Two Pointers/M-1456. Maximum Number of Vowels in a Substring of Given Length/M-1456. Maximum Number of Vowels in a Substring of Given Length.py
class Solution:
    @classmethod
    def maxVowels(self, s: str, k: int) -> int:
        vowels_cnt, max_vowels_cnt = 0, 0
        vowels = {"a", "e", "i", "o", "u"}
        l, r = 0, 0

        while r < len(s):
            if s[r] in vowels:
                vowels_cnt += 1
                max_vowels_cnt = max(max_vowels_cnt, vowels_cnt)

            # reach the window size
            if r - l == k - 1:
                if s[l] in vowels:
                    vowels_cnt -= 1
                l += 1

            r += 1

        return max_vowels_cnt


testData = [
    {
        "s": "abciiidef",
        "k": 3,
        "ans": 3,
    },
    {
        "s": "aeiou",
        "k": 2,
        "ans": 2,
    },
    {
        "s": "leetcode",
        "k": 3,
        "ans": 2,
    },
]

for data in testData:
    print(data["ans"] == Solution.maxVowels(data["s"], data["k"]))
